Propagating long times does not converge

<!--- Provide a general summary of the issue in the Title above -->

🐞 **Problem**

Propagating long times fails in different ways:

```python
In [5]: iss.propagate(100 * u.year)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-27-503eb53e62e7> in <module>()
----> 1 iss.propagate(100 * u.year)

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in propagate(self, epoch_or_duration, rtol)
    245             time_of_flight = time.TimeDelta(epoch_or_duration)
    246 
--> 247         return propagate(self, time_of_flight, rtol=rtol)
    248 
    249     def sample(self, *, num_points=None, time_values=None):

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in propagate(orbit, time_of_flight, method, rtol, **kwargs)
    156                   time_of_flight.to(u.s).value,
    157                   rtol=rtol,
--> 158                   **kwargs)
    159     return orbit.from_vectors(orbit.attractor, r * u.km, v * u.km / u.s, orbit.epoch + time_of_flight)
    160 

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in kepler(k, r0, v0, tof, rtol, numiter)
    139     f, g, fdot, gdot = _kepler(k, r0, v0, tof, numiter, rtol)
    140 
--> 141     assert np.abs(f * gdot - fdot * g - 1) < 1e-5  # Fixed tolerance
    142 
    143     # Return position and velocity vectors

AssertionError: 

In [6]: halleys[0].propagate(73 * u.year)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-26-9dcacad8e904> in <module>()
----> 1 halleys[0].propagate(73 * u.year)

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/orbit.py in propagate(self, epoch_or_duration, rtol)
    245             time_of_flight = time.TimeDelta(epoch_or_duration)
    246 
--> 247         return propagate(self, time_of_flight, rtol=rtol)
    248 
    249     def sample(self, *, num_points=None, time_values=None):

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in propagate(orbit, time_of_flight, method, rtol, **kwargs)
    156                   time_of_flight.to(u.s).value,
    157                   rtol=rtol,
--> 158                   **kwargs)
    159     return orbit.from_vectors(orbit.attractor, r * u.km, v * u.km / u.s, orbit.epoch + time_of_flight)
    160 

~/Personal/poliastro/poliastro-library/src/poliastro/twobody/propagation.py in kepler(k, r0, v0, tof, rtol, numiter)
    137     """
    138     # Compute Lagrange coefficients
--> 139     f, g, fdot, gdot = _kepler(k, r0, v0, tof, numiter, rtol)
    140 
    141     assert np.abs(f * gdot - fdot * g - 1) < 1e-5  # Fixed tolerance

RuntimeError: Maximum number of iterations reached
```

🎯 **Goal**

One thing is that the long term errors of propagation are big. But nonetheless, the propagator should always give an answer, instead of an internal error (at least for a full period, see the failure of the Halley's comet). I wonder if this is possible at all?

💡 **Possible solutions**

The propagator of poliastro uses the universal variables approach, which is popular but has several limitations. Perhaps for long propagation times it should be replaced with a Newton-Raphson iteration and a good starting point as suggested by http://dx.doi.org/10.1007/s10569-013-9476-9.

📋  **Steps to solve the problem**

 * Comment below about what you've started working on.
 * Add, commit, push your changes
 * Submit a pull request and add this in comments - `Addresses #<put issue number here>`
 * Ask for a review in comments section of pull request
 * Celebrate your contribution to this project 🎉

--- a/src/poliastro/tests/tests_twobody/test_propagation.py
+++ b/src/poliastro/tests/tests_twobody/test_propagation.py
@@ -12,7 +12,8 @@ from poliastro.twobody.rv import rv2coe
 from poliastro.constants import J2000
 from poliastro.bodies import Sun, Earth
 from poliastro.twobody import Orbit
-from poliastro.twobody.propagation import cowell, kepler, mean_motion, cowell
+from poliastro.twobody.propagation import cowell, kepler, mean_motion
+from poliastro.examples import iss
 
 from poliastro.util import norm
 
@@ -206,3 +207,25 @@ def test_propagate_to_date_has_proper_epoch():
 
     # Tolerance should be higher, see https://github.com/astropy/astropy/issues/6638
     assert (ss1.epoch - final_epoch).sec == approx(0.0, abs=1e-6)
+
+
+@pytest.mark.filterwarnings('ignore:ERFA')
+@pytest.mark.filterwarnings('ignore::UserWarning')
+@pytest.mark.parametrize('method', [
+    mean_motion,
+    pytest.param(cowell, marks=pytest.mark.xfail),
+    pytest.param(kepler, marks=pytest.mark.xfail),
+])
+def test_propagate_long_times_keeps_geometry(method):
+    # See https://github.com/poliastro/poliastro/issues/265
+    time_of_flight = 100 * u.year
+
+    res = iss.propagate(time_of_flight, method=method)
+
+    assert_quantity_allclose(iss.a, res.a)
+    assert_quantity_allclose(iss.ecc, res.ecc)
+    assert_quantity_allclose(iss.inc, res.inc)
+    assert_quantity_allclose(iss.raan, res.raan)
+    assert_quantity_allclose(iss.argp, res.argp)
+
+    assert_quantity_allclose((res.epoch - iss.epoch).to(time_of_flight.unit), time_of_flight)

--- a/src/poliastro/tests/tests_twobody/test_sample.py
+++ b/src/poliastro/tests/tests_twobody/test_sample.py
@@ -4,7 +4,7 @@ from astropy.tests.helper import assert_quantity_allclose
 from astropy import units as u
 from astropy.time import Time
 
-from poliastro.bodies import Earth
+from poliastro.bodies import Earth, Sun
 from poliastro.twobody import Orbit
 from poliastro.twobody.propagation import kepler, mean_motion, cowell
 import numpy as np
@@ -91,3 +91,21 @@ def test_sample_num_points(num_points):
 
     assert len(rr) == num_points
     # assert_quantity_allclose(rr[num_points // 2].get_xyz(), expected_ss.r)
+
+
+@pytest.mark.parametrize('method', [
+    mean_motion,
+    cowell,
+    pytest.param(kepler, marks=pytest.mark.xfail),
+])
+def test_sample_big_orbits(method):
+    # See https://github.com/poliastro/poliastro/issues/265
+    ss = Orbit.from_vectors(
+        Sun,
+        [-9018878.6, -94116055, 22619059] * u.km,
+        [-49.950923, -12.948431, -4.2925158] * u.km / u.s
+    )
+
+    times, positions = ss.sample(15, method=method)
+
+    assert len(times) == len(positions) == 15

--- a/src/poliastro/twobody/propagation.py
+++ b/src/poliastro/twobody/propagation.py
@@ -107,7 +107,7 @@ def cowell(k, r0, v0, tof, rtol=1e-10, *, ad=None, callback=None, nsteps=1000):
 
 
 def mean_motion(k, r0, v0, tof, **kwargs):
-    """Propagates orbit using mean motion
+    r"""Propagates orbit using mean motion
 
     Parameters
     ----------
@@ -122,11 +122,12 @@ def mean_motion(k, r0, v0, tof, **kwargs):
     tof : float
         Time of flight (s).
 
-    Note
+    Notes
     -----
-    This method takes initial \vec{r}, \vec{v}, calculates classical orbit parameters,
-    increases mean anomaly and performs inverse transformation to get final \vec{r}, \vec{v}
+    This method takes initial :math:`\vec{r}, \vec{v}`, calculates classical orbit parameters,
+    increases mean anomaly and performs inverse transformation to get final :math:`\vec{r}, \vec{v}`
     The logic is based on formulae (4), (6) and (7) from http://dx.doi.org/10.1007/s10569-013-9476-9
+
     """
 
     # get the initial true anomaly and orbit parameters that are constant over time

